ako-crd-operator/internal/controller/applicationprofile_controller.go

```diff
@@ -133,26 +141,40 @@ func (r *ApplicationProfileReconciler) SetupWithManager(mgr ctrl.Manager) error
 		Complete(r)
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *ApplicationProfileReconciler) DeleteObject(ctx context.Context, ap *akov1alpha1.ApplicationProfile) error {
+// DeleteObject deletes the ApplicationProfile from Avi Controller and returns (error, bool)
+// The boolean indicates whether the finalizer should be removed (true) or kept (false)
+func (r *ApplicationProfileReconciler) DeleteObject(ctx context.Context, ap *akov1alpha1.ApplicationProfile) (error, bool) {
 	log := utils.LoggerFromContext(ctx)
 	if ap.Status.UUID != "" {
 		if err := r.AviClient.AviSessionDelete(utils.GetUriEncoded(fmt.Sprintf("%s/%s", constants.ApplicationProfileURL, ap.Status.UUID)), nil, nil); err != nil {
-
 			// Handle 404 as success case - object doesn't exist, which is the desired state for delete
-			if aviError, ok := err.(session.AviError); ok && aviError.HttpStatusCode == 404 {
-				log.Info("ApplicationProfile not found on Avi Controller (404), treating as successful deletion")
-				return nil
+			if aviError, ok := err.(session.AviError); ok {
+				switch aviError.HttpStatusCode {
+				case 404:
+					log.Info("ApplicationProfile not found on Avi Controller (404), treating as successful deletion")
+					return nil, true
+				case 403:
+					log.Errorf("ApplicationProfile is being referred by other objects, cannot be deleted. %s", aviError.Error())
+					r.EventRecorder.Event(ap, corev1.EventTypeWarning, "DeletionSkipped", aviError.Error())
+					ap.Status.Conditions = controllerutils.SetCondition(ap.Status.Conditions, metav1.Condition{
+						Type:               "Delete",
```

Comment:
NIT: Type can be "Deleted" . False status would then mean resource is not deleted.  since metaV1.ConditionFalse means a resource is not in the condition.